utils/gear-tasks/src/host.rs

```diff
@@ -0,0 +1,215 @@
+    pub(crate) struct GearTasksContextExt;
+}
+
+/// The tasks' runner we spawn after node client creation.
+pub struct GearTasksRunner<RA, Block> {
+    runtime_api_provider: Arc<RA>,
+    rx: mpsc::Receiver<TasksRunnerEvent>,
+    thread_pool: Option<ThreadPool>,
+    _block: PhantomData<Block>,
+}
+
+impl<RA, Block> GearTasksRunner<RA, Block>
+where
+    RA: ProvideRuntimeApi<Block> + UsageProvider<Block> + Send + Sync + 'static,
+    RA::Api: GearTasksApi<Block> + RuntimeSetOverlayedChanges<Block>,
+    Block: BlockT,
+{
+    pub fn new(client: Arc<RA>) -> Self {
+        let (tx, rx) = mpsc::channel();
+        assert!(
+            RUNNER_TX.get().is_none(),
+            "`GearTasksRunner` initialized twice"
+        );
+        let _tx = RUNNER_TX.get_or_init(move || tx);
+
+        Self {
+            runtime_api_provider: client,
+            rx,
+            thread_pool: None,
+            _block: PhantomData,
+        }
+    }
+
+    pub async fn run(mut self) {
+        for info in self.rx {
+            match info {
+                TasksRunnerEvent::ReInit { tasks } => {
+                    self.thread_pool = Some(
+                        ThreadPool::builder()
+                            .pool_size(tasks as usize)
+                            .name_prefix("gear-tasks-")
+                            .create()
+                            .expect("Thread pool creation failed"),
+                    );
+                }
+                TasksRunnerEvent::Spawn {
+                    overlayed_changes,
+                    func_ref,
+                    payload,
+                    rx,
+                } => {
+                    let client = self.runtime_api_provider.clone();
+                    let thread_pool = self
+                        .thread_pool
+                        .as_ref()
+                        .expect("`TasksRunnerEvent::ReInit` has never been sent");
+                    thread_pool.spawn_ok(async move {
+                        let mut runtime_api = client.runtime_api();
+                        runtime_api.register_extension(GearTasksContextExt);
+
+                        let mut overlayed_changes = overlayed_changes
+                            .downcast::<OverlayedChanges<HashingFor<Block>>>()
+                            .expect(
+                                "`Externalities::gear_overlayed_changes()` returns invalid type",
+                            );
+                        // Overlayed changes enter runtime mode in `StateMachine::execute()`
+                        // so we definitely want to exit from it before `execute_task()` tries to
+                        // enter it again.
+                        overlayed_changes.exit_runtime().expect(
+                            "Overlayed changes is never called in `sp_state_machine::StateMachine::execute()`",
+                        );
+                        runtime_api.set_overlayed_changes(*overlayed_changes);
```

Comment:
if we can ensure this logic runs before state machine execution, why we still need to reset the overlayed_changes of state machine (it's holding a mutable reference of OverlayedChanges) but not modify it with `RuntimeApiImpl.changes` directly?

see also: [impl CallExecutor for sc_service::LocalExecutor](https://github.com/gear-tech/polkadot-sdk/blob/master/substrate/client/service/src/client/call_executor.rs#L148)